library/core/src/slice/mod.rs

```diff
@@ -1843,25 +1843,24 @@ impl<T> [T] {
     where
         F: FnMut(&'a T) -> Ordering,
     {
-        let s = self;
-        let mut size = s.len();
-        if size == 0 {
-            return Err(0);
-        }
-        let mut base = 0usize;
-        while size > 1 {
-            let half = size / 2;
-            let mid = base + half;
-            // SAFETY: the call is made safe by the following inconstants:
-            // - `mid >= 0`: by definition
-            // - `mid < size`: `mid = size / 2 + size / 4 + size / 8 ...`
-            let cmp = f(unsafe { s.get_unchecked(mid) });
-            base = if cmp == Greater { base } else { mid };
-            size -= half;
+        let mut left = 0;
+        let mut right = self.len();
+        while left < right {
+            // never overflow because `slice::len()` max is `isize::MAX`.
+            let mid = (left + right) / 2;
+            // SAFETY: the call is made safe by the following invariants:
+            // - `mid >= 0`
+            // - `mid < size`: `mid` is limited by `[left; right)` bound.
+            let cmp = f(unsafe { self.get_unchecked(mid) });
+            if cmp == Less {
```

Comment:
Is there a particular reason not to use `match`?

Reply:
Yeah, `match` reorder comparison operations, which is perf sensitive.
This is x86 asm for u8: https://rust.godbolt.org/z/8Y8Pra

Reply:
Indeed the code generated for `match` is terrible. We'd better have a comment here explaining why if's are used instead.